pkg/controller/metrics/metrics.go

```diff
@@ -197,6 +224,35 @@ type Calculator struct {
 func (mc *Calculator) Start(ctx context.Context) error {
 	log.Info("started metrics calculator goroutine")
 
+	// Register optional metrics and add them to corresponding maps, so controllers logging them can access
+	// the information
+	hiveConfig := &hivev1.HiveConfig{}
+	err := mc.Client.Get(context.TODO(), types.NamespacedName{Name: constants.HiveConfigName}, hiveConfig)
+	if err == nil && hiveConfig.Spec.MetricsConfig != nil {
+		if durationMetrics := hiveConfig.Spec.MetricsConfig.MetricsWithDuration; durationMetrics != nil {
+			MapMetricToDuration = make(map[*prometheus.HistogramVec]time.Duration)
+			for _, metric := range durationMetrics {
+				switch metric.Name {
+				case hivev1.CurrentStopping:
+					metrics.Registry.MustRegister(MetricStoppingClusters)
+					MapMetricToDuration[MetricStoppingClusters] = metric.Duration.Duration
+				case hivev1.CurrentResuming:
+					metrics.Registry.MustRegister(MetricResumingClusters)
+					MapMetricToDuration[MetricResumingClusters] = metric.Duration.Duration
+				case hivev1.CurrentWaitingForCO:
+					metrics.Registry.MustRegister(MetricWaitingForCOClusters)
+					MapMetricToDuration[MetricWaitingForCOClusters] = metric.Duration.Duration
+				case hivev1.CumulativeHibernated:
+					metrics.Registry.MustRegister(MetricClusterHibernationTransitionSeconds)
+					MapMetricToDuration[MetricClusterHibernationTransitionSeconds] = metric.Duration.Duration
+				case hivev1.CumulativeResumed:
+					metrics.Registry.MustRegister(MetricClusterReadyTransitionSeconds)
+					MapMetricToDuration[MetricClusterReadyTransitionSeconds] = metric.Duration.Duration
```

Comment:
I tried to have a direct mapping with hiveconfig types, but golang's strict rules won't allow me to dynamically register a metric

Reply:
Not following this comment. Let's discuss.